README.md
# Interactive Solid authentication in Node.js

[![Build status](https://github.com/rubensworks/solid-node-interactive-auth.js/workflows/CI/badge.svg)](https://github.com/rubensworks/solid-node-interactive-auth.js/actions?query=workflow%3ACI)
[![Coverage Status](https://coveralls.io/repos/github/rubensworks/solid-node-interactive-auth.js/badge.svg?branch=master)](https://coveralls.io/github/rubensworks/solid-node-interactive-auth.js?branch=master)
[![npm version](https://badge.fury.io/js/solid-node-interactive-auth.svg)](https://www.npmjs.com/package/solid-node-interactive-auth)

Easily authenticate Node.js apps with Solid identity servers by opening the user's Web browser.

Internally, this tool will setup a temporary Web server on the localhost
to allow authentication data to be handled easily and safely.

This is to be used as a tool next to [`@inrupt/solid-client-authn-node`](https://www.npmjs.com/package/@inrupt/solid-client-authn-node).

## Try out how it works

To see how the interactive authentication works for the end-user,
you can run this command:

```bash
$ npx solid-node-interactive-auth https://solidcommunity.net/

Logged in as https://<MY USERNAME>.solidcommunity.net/profile/card#me
```

You can replace https://solidcommunity.net/ with the identity provider you want to authenticate with.

## Installation

```bash
$ npm install solid-node-interactive-auth
```
or
```bash
$ yarn add solid-node-interactive-auth
```

This tool requires [`@inrupt/solid-client-authn-node`](https://www.npmjs.com/package/@inrupt/solid-client-authn-node) as a peer dependency:

```bash
$ npm install @inrupt/solid-client-authn-node
```
or
```bash
$ yarn add @inrupt/solid-client-authn-node
```

## Usage

The following code will trigger the user's Web browser to be opened to trigger the login sequence:
```typescript
import { Session } from '@inrupt/solid-client-authn-node';
import { login } from 'solid-node-interactive-auth';

(async function() {
  // Create a new session and log in by opening the Web browser
  const session = new Session();
  await login({
    session,
    oidcIssuer: 'https://solidcommunity.net/',
    tokenType: 'Bearer',
  });

  // Perform operations with this session
  // such as session.fetch()

  // Log out once you're done (avoids hanging Node.js process)
  await session.logout();
})();
```

If you don't have any specific needs for the `Session` object,
you can also just let this tool create one for you:
```typescript
const session = await login({
  oidcIssuer: 'https://solidcommunity.net/',
  tokenType: 'Bearer',
});
```

## License
This software is written by [Ruben Taelman](http://rubensworks.net/).

This code is released under the [MIT license](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

<a name="v1.1.1"></a>
## [v1.1.1](https://github.com/rubensworks/solid-node-interactive-auth.js/compare/v1.1.0...v1.1.1) - 2022-11-09

### Fixed
* [Include source map files in packed files](https://github.com/rubensworks/solid-node-interactive-auth.js/commit/34498c9c8e53df945d7003dc3534da442e2efeda)

<a name="v1.1.0"></a>
## [v1.1.0](https://github.com/rubensworks/solid-node-interactive-auth.js/compare/v1.0.2...v1.1.0) - 2022-10-31

### Changed
* [Clear code param after authentication](https://github.com/rubensworks/solid-node-interactive-auth.js/commit/40603c512b7971103dcf8a8a6645948f8d5a346f)

<a name="v1.0.2"></a>
## [v1.0.2](https://github.com/rubensworks/solid-node-interactive-auth.js/compare/v1.0.1...v1.0.2) - 2021-08-23

### Fixed
* [Don't default to Bearer tokens, as they don't work on all IDPs](https://github.com/rubensworks/solid-node-interactive-auth.js/commit/2c26611a8a166d30471264e4fa4324b3350c5e9a)

<a name="v1.0.1"></a>
## [v1.0.1](https://github.com/rubensworks/solid-node-interactive-auth.js/compare/v1.0.0...v1.0.1) - 2021-08-23

### Fixed
* [Fix missing bin entry in package.json](https://github.com/rubensworks/solid-node-interactive-auth.js/commit/215f98d87b26fa9c48505cc71b5206a5802f0ac5)

<a name="v1.0.0"></a>
## [v1.0.0] - 2021-08-23

Initial release


test/HttpLoginHandler.test.ts
import type { Server, ServerResponse } from 'http';
import type * as net from 'net';
import type { Session } from '@inrupt/solid-client-authn-node';
import * as open from 'open';
import { HttpLoginHandler } from '../lib/HttpLoginHandler';

let server: Server;
let requestHandler: any;
jest.mock('http', () => ({
  createServer(requestHandlerThis: any) {
    requestHandler = requestHandlerThis;
    return server;
  },
}));

jest.mock('open', () => jest.fn());

describe('HttpLoginHandler', () => {
  let handler: HttpLoginHandler;
  let socket: net.Socket;
  let session: Session;
  let res: ServerResponse;
  beforeEach(() => {
    jest.clearAllMocks();

    res = <any> {
      writeHead: jest.fn(),
      end: jest.fn(),
    };
    socket = <any> {
      destroy: jest.fn(),
    };
    server = <any> {
      listen: jest.fn(),
      on(event: string, cb: any) {
        if (event === 'connection') {
          cb(socket);
        }
      },
      close: jest.fn(),
    };
    session = <any> {
      handleIncomingRedirect: jest.fn(async() => {
        // Do nothing
      }),
      login: jest.fn((options: any) => {
        options.handleRedirect('http://example.org/authenticate');

        requestHandler({
          url: '/onLoggedIn',
        }, res);

        requestHandler({
          url: '/done',
        }, res);
      }),
      info: {
        isLoggedIn: true,
      },
    };
    handler = new HttpLoginHandler({
      port: 3005,
      session,
    });
  });

  describe('handleLogin', () => {
    it('handles a valid login sequence', async() => {
      await handler.handleLogin({});

      expect(server.listen).toHaveBeenCalledWith(3005);

      expect(session.login).toHaveBeenCalledWith({
        clientName: '@rubensworks/solid-node-interactive-auth',
        handleRedirect: expect.any(Function),
        redirectUrl: 'http://localhost:3005/onLoggedIn',
      });
      expect(open).toHaveBeenCalledWith('http://example.org/authenticate');
      expect(session.handleIncomingRedirect).toHaveBeenCalledWith('http://localhost:3005/onLoggedIn');
      expect(res.writeHead).toHaveBeenCalledWith(200);
      expect(res.end).toHaveBeenCalledWith(
        '<script>window.location = new URL(\'http://localhost:3005/done\'); window.close();</script>',
      );

      expect(server.close).toHaveBeenCalledWith();
      expect(socket.destroy).toHaveBeenCalledWith();

      await open('http://localhost:3005/done');
    });

    it('handles a valid post-login sequence', async() => {
      await handler.handleLogin({});

      await open('http://localhost:3005/done');
      expect(res.writeHead).toHaveBeenCalledWith(200);
      expect(res.end).toHaveBeenCalledWith(
        '<script>window.close();</script>',
      );

      expect(server.close).toHaveBeenCalledWith();
      expect(socket.destroy).toHaveBeenCalledWith();
    });

    it('handles a valid login sequence with custom options', async() => {
      const handleRedirect = jest.fn();
      await handler.handleLogin({
        clientName: 'OTHER CLIENT',
        handleRedirect,
      });

      expect(server.listen).toHaveBeenCalledWith(3005);

      expect(session.login).toHaveBeenCalledWith({
        clientName: 'OTHER CLIENT',
        handleRedirect,
        redirectUrl: 'http://localhost:3005/onLoggedIn',
      });
      expect(open).not.toHaveBeenCalled();
      expect(handleRedirect).toHaveBeenCalledWith('http://example.org/authenticate');
      expect(session.handleIncomingRedirect).toHaveBeenCalledWith('http://localhost:3005/onLoggedIn');
      expect(res.writeHead).toHaveBeenCalledWith(200);
      expect(res.end).toHaveBeenCalledWith(
        '<script>window.location = new URL(\'http://localhost:3005/done\'); window.close();</script>',
      );

      expect(server.close).toHaveBeenCalledWith();
      expect(socket.destroy).toHaveBeenCalledWith();
    });

    it('returns 404 on invalid requests', async() => {
      await handler.handleLogin({});

      requestHandler({
        url: '/invalid',
      }, res);

      expect(res.writeHead).toHaveBeenCalledWith(404);
      expect(res.end).toHaveBeenCalledWith('This temporary server only accepts /onLoggedIn requests');
    });

    it('rejects when login has failed', async() => {
      session.info.isLoggedIn = false;
      await expect(handler.handleLogin({})).rejects.toThrowError('Login process has failed');
    });
  });
});


lib/HttpLoginHandler.ts
import * as http from 'http';
import type { Server } from 'http';
import type * as net from 'net';
import type { ILoginInputOptions } from '@inrupt/solid-client-authn-core';
import type { Session } from '@inrupt/solid-client-authn-node';
import * as open from 'open';

/**
 * Handles the Solid login process by starting a temporary HTTP server.
 */
export class HttpLoginHandler {
  private readonly port: number;
  private readonly session: Session;

  public constructor(options: IHttpLoginHandlerOptions) {
    this.port = options.port;
    this.session = options.session;
  }

  /**
   * Start a temporary HTTP server for handling the login process.
   * @param options Login options.
   */
  public async handleLogin(options: ILoginInputOptions): Promise<void> {
    // Determine login options
    options.redirectUrl = `http://localhost:${this.port}/onLoggedIn`;
    if (!options.clientName) {
      options.clientName = '@rubensworks/solid-node-interactive-auth';
    }
    if (!options.handleRedirect) {
      options.handleRedirect = redirectUrl => open(redirectUrl);
    }

    // Start temporary server
    let server: Server;
    const sockets: net.Socket[] = [];
    const onLoggedIn = new Promise((resolve, reject) => {
      server = http.createServer((req, res) => {
        if (req.url!.startsWith('/onLoggedIn')) {
          this.session.handleIncomingRedirect(`http://localhost:${this.port}${req.url}`)
            .then(resolve, reject);
          res.writeHead(200);
          res.end(`<script>window.location = new URL('http://localhost:${this.port}/done'); window.close();</script>`);
        } else if (req.url!.startsWith('/done')) {
          res.writeHead(200);
          res.end(`<script>window.close();</script>`);
        } else {
          res.writeHead(404);
          res.end(`This temporary server only accepts /onLoggedIn requests`);
        }
      });
      server.listen(this.port);

      // Avoid hanging server after close
      server.on('connection', socket => sockets.push(socket));
    });

    // Trigger login process
    await this.session.login(options);

    // Wait for server to handle incoming redirect
    await onLoggedIn;

    // Close the server
    server!.close();
    for (const socket of sockets) {
      socket.destroy();
    }

    // Ensure we are logged in
    if (!this.session.info.isLoggedIn) {
      throw new Error(`Login process has failed`);
    }
  }
}

export interface IHttpLoginHandlerOptions {
  /**
   * Port of the temporary HTTP server to start on the local host.
   */
  port: number;
  /**
   * The session to authenticate.
   */
  session: Session;
}
